internal/controller/linodefirewall_controller.go

```diff
@@ -205,23 +207,47 @@ func (r *LinodeFirewallReconciler) reconcile(
 	return ctrl.Result{}, nil
 }
 
-func (r *LinodeFirewallReconciler) removeAddressSetFinalizer(ctx context.Context, logger logr.Logger, fwScope *scope.FirewallScope, addrSetRef *corev1.ObjectReference) error {
-	if addrSetRef == nil || r.Client == nil {
+func (r *LinodeFirewallReconciler) removeAddressSetFinalizer(ctx context.Context, logger logr.Logger, fwScope *scope.FirewallScope, objRef *corev1.ObjectReference) error {
+	if objRef == nil || r.Client == nil {
 		return nil
 	}
 	addrSet := &infrav1alpha2.AddressSet{}
-	if addrSetRef.Namespace == "" {
-		addrSetRef.Namespace = fwScope.LinodeFirewall.Namespace
+	if objRef.Namespace == "" {
+		objRef.Namespace = fwScope.LinodeFirewall.Namespace
 	}
-	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: addrSetRef.Namespace, Name: addrSetRef.Name}, addrSet); err != nil {
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: objRef.Namespace, Name: objRef.Name}, addrSet); err != nil {
 		if !apierrors.IsNotFound(err) {
-			logger.Error(err, "failed to fetch referenced AddressSet", "namespace", addrSetRef.Namespace, "name", addrSetRef.Name)
+			logger.Error(err, "failed to fetch referenced Object", "kind", objRef.Kind, "namespace", objRef.Namespace, "name", objRef.Name)
 			return err
 		}
 	}
 	if controllerutil.RemoveFinalizer(addrSet, getFinalizer(fwScope.LinodeFirewall)) {
 		if err := r.Update(ctx, addrSet); err != nil {
-			logger.Error(err, "failed to remove finalizer from AddressSet")
+			logger.Error(err, "failed to remove finalizer from Object", "kind", objRef.Kind, "namespace", objRef.Namespace, "name", objRef.Name)
```

Comment:
Does this create a race condition when a user tries to delete multiple `LinodeFirewall`s that reference the same `AddressSet` simultaneously?

Reply:
Added the retry package for both adding and removing finalizers for firewallrules and addressets 👍